### PR TITLE
clear quick fix decoration appropriately

### LINKED ISSUE #162180: After triggering the create PR quick fix we should remove the lightbulb as it's fixed
Repro:

- Push a new branch to a GH repo
- Trigger the quick fix to open the PR link
- The lightbulb remains

image

Ideally we would remove just that quick fix from the list as it's actioned (if there are multiple), or just hide the lightbulb completely.

### LINKED ISSUE #162216: Quickfix lightbulb appends to any existing command after applying
Testing #161846

1. Do something that prompts a quick fix
2. Start typing a command, then realize there's a suggested fix for it
3. Accept the quickfix suggestion
4. Notice the last command is appended to whatever you had typed in step 2

Could be fixed by:

- Hiding the bulb if anything new is typed on the prompt
- Clearing the prompt when applying the fix
- Or just not appending the last command in the case something is already there

### PR DESCRIPTION
**Pull request by meganrogge**
fix https://github.com/microsoft/vscode/issues/162180
fix [#162216](https://github.com/microsoft/vscode/issues/162216)

### CODE REVIEW

**Review by Tyriar — COMMENTED** [review]
Is https://github.com/microsoft/vscode/issues/162216 resolved in this? I was expecting a refactor to pull this out and use it:

https://github.com/microsoft/vscode/blob/132b81c1970d264f466b17f44b4d6fe8ca67c9ef/src/vs/workbench/contrib/terminal/browser/terminalRunRecentQuickPick.ts#L287-L299

**Comment by meganrogge** [comment]
@Tyriar yes good pt, that's the next part I need to do. What's done thus far in this PR is managing a map of decorations/quick fixes instead of just one such that we can remove actions as they get run and won't remove the decoration until the quick fix list is empty

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 1]
```diff
@@ -916,6 +916,8 @@ export interface ITerminalInstance {
 	registerQuickFixProvider(...options: ITerminalQuickFixOptions[]): void;
 
 	freePortKillProcess(port: string): Promise<void>;
+
+	runCommand(command: string, alt?: boolean): void;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:920]: Let's move this to just after sendText/sendPath as it's closely related, also adding docs and renaming alt to something more descriptive like `addNewLine?: boolean` or `executeCommand?: boolean` would be good.

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 2]
```diff
@@ -94,70 +95,111 @@ export class TerminalQuickFixAddon extends Disposable implements ITerminalAddon,
...
+	 */
+	private _registerQuickFixDecoration(): boolean {
+		if (!this._terminal) {
+			return false;
+		}
+		const id = this._quickFixId;
+		if (!id || !this._quickFixes.get(id) || this._decorations.get(id)) {
+			return false;
+		}
 		const marker = this._terminal.registerMarker();
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:166]: The decoration may not be positioned correctly after enter:

image

Might only happen on Windows after the repositioning of the decoration that can occur.

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 3]
```diff
@@ -46,15 +48,13 @@ export class TerminalQuickFixAddon extends Disposable implements ITerminalAddon,
...
 
-	private _currentQuickFixElement: HTMLElement | undefined;
-
-	private _decorationMarkerIds = new Set<number>();
-
 	private _commandListeners: Map<string, ITerminalQuickFixOptions[]> = new Map();
 
-	private _quickFixes: IAction[] | undefined;
+	private _quickFixes: Map<number, IAction[]> = new Map();
+	private _decorations: Map<number, IDecoration> = new Map();
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:54]: I don't think we need the maps if we're doing it the simpler way?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:54]: we do and it's not totally clear why - spent awhile yesterday looking into it and didn't work at all when I went back to just a single decoration

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:166]: I'm not convinced that we should be repositioning the decoration at all. It adds complexity like this and doesn't add much benefit considering the shortcut will still work as will clicking on it several lines above

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts [hunk 4]
```diff
@@ -94,70 +95,111 @@ export class TerminalQuickFixAddon extends Disposable implements ITerminalAddon,
...
-export function getQuickFixes(
+export function getQuickFixesForCommand(
 	command: ITerminalCommand,
 	actionOptions: Map<string, ITerminalQuickFixOptions[]>,
 	openerService: IOpenerService,
 	onDidRequestRerunCommand?: Emitter<{ command: string; addNewLine?: boolean }>
-): IAction[] | undefined {
-	const actions: IAction[] = [];
+): { fixes: IAction[]; onDidRunQuickFix: Event<void> } | undefined {
+	const _onDidRunQuickFix = new Emitter<void>();
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:200]: Lint error:

```
Variable name `_onDidRunQuickFix` must match one of the following formats: camelCase, UPPER_CASE, PascalCaseeslint[@typescript-eslint/naming-convention](https://typescript-eslint.io/rules/naming-convention)
```

Change to `onDidRunQuickFixEmitter`?

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts (hunk 2 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:166]: That's fair. In that case, could we still keep it on top of the previous placeholder? It moves atm:

Recording 2022-10-06 at 07 28 55

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts (hunk 3 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:54]: We only need to track a single list of quick fixes now though right? Same for a single decoration? Removing the maps should greatly simplify things.

File: src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/browser/xterm/quickFixAddon.ts:166]: I'll just undo the changes that make it move

**Review by Tyriar — CHANGES_REQUESTED** [review]
lgtm after the lint rule (no variable underscore prefix) is happy

**Review by Tyriar — APPROVED** [review]

**Pull request merged**